File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.cpp
class Solution {
    int partition(vector<int>& nums, int left, int right) {
        int pivot = nums[left], l = left + 1, r = right;
        while (l <= r) {
            if (nums[l] < pivot && nums[r] > pivot) 
                swap(nums[l++], nums[r--]);
            if (nums[l] >= pivot) 
                l++;
            if (nums[r] <= pivot)
                r--;
        }
        swap(nums[left], nums[r]);
        return r;
    }
public:
    int findKthLargest(vector<int>& nums, int k) {
        int left = 0, right = nums.size() - 1;
        while (true) {
            int idx = partition(nums, left, right);
            if (idx == k - 1) 
                return nums[idx];
            if (idx < k - 1) 
                left = idx + 1;
            else 
                right = idx - 1;
        }
    }
};

File: 215-kth-largest-element-in-an-array/215-kth-largest-element-in-an-array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        
        length = len(nums)
        
        def partition(index, end):
            left = index + 1
            right = end - 1
            while left <= right:
                if nums[right] >= nums[index]:
                    nums[right], nums[left] = nums[left], nums[right]
                    left += 1
                else:
                    right -= 1
            nums[index], nums[right] = nums[right], nums[index]
            return right
        
        
        start = 0
        end = length
        k = k - 1
        while True:
            index = partition(start, end)
            if index == k:
                return nums[index]
            if index > k:
                end = index
            else:
                start = index + 1
            
        return 0